pkg/timeutil/timespan.go

```diff
@@ -0,0 +1,186 @@
+package timeutil
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var (
+	badDay = time.Weekday(-1)
+
+	shortDayNames = map[string]time.Weekday{
+		"Sun": time.Sunday,
+		"Mon": time.Monday,
+		"Tue": time.Tuesday,
+		"Wed": time.Wednesday,
+		"Thu": time.Thursday,
+		"Fri": time.Friday,
+		"Sat": time.Saturday,
+	}
+)
+
+func d2s(d time.Weekday) string {
+	for k, v := range shortDayNames {
+		if d == v {
+			return k
+		}
+	}
+
+	return ""
+}
+
+type dayTime struct {
+	day     time.Weekday // day of week, -1 if unset
+	minutes int          // minutes into day
+}
+
+func parseClock(s string) (hh, mm int, err error) {
+	spl := strings.Split(s, ":")
+	if len(spl) != 2 {
+		err = errors.New("malformed HH:MM specification")
+		return
+	}
+
+	hh, err = strconv.Atoi(spl[0])
+	if err != nil {
+		err = fmt.Errorf("malformed hour %q: %s", spl[0], err)
+		return
+	}
+
+	mm, err = strconv.Atoi(spl[1])
+	if err != nil {
+		err = fmt.Errorf("malformed minute %q: %s", spl[1], err)
+		return
+	}
+
+	return
+}
+
+func parsedaytime(s string) (dayTime, error) {
+	dt := dayTime{badDay, 0}
+	spl := strings.Fields(s)
+
+	switch len(spl) {
+	default:
+		fallthrough
+	case 0:
+		return dt, errors.New("empty time specification")
+	case 2:
+		// 2 fields, day hh:mm
+		for s, wd := range shortDayNames {
+			if strings.ToLower(spl[0]) == strings.ToLower(s) {
+				dt.day = wd
+			}
+		}
+
+		if dt.day == time.Weekday(-1) {
+			return dt, fmt.Errorf("invalid weekday %q", spl[0])
+		}
+
+		spl = spl[1:]
+		fallthrough
+	case 1:
+		// 1 field, must be time.
+		hh, mm, err := parseClock(spl[0])
+		if err != nil {
+			return dt, err
+		}
+
+		dt.minutes = (hh * 60) + mm
+		return dt, nil
+	}
+}
+
+type Window struct {
+	start dayTime
+	end   dayTime
+}
+
+func ParseWindow(start string, end string) (*Window, error) {
+	s, err := parsedaytime(start)
+	if err != nil {
+		return nil, fmt.Errorf("start time %q: %s", start, err)
+	}
+
+	e, err := parsedaytime(end)
+	if err != nil {
+		return nil, fmt.Errorf("end time %q: %s", end, err)
+	}
+
+	// check weekday is present in both or neither
+	if (s.day == badDay) != (e.day == badDay) {
```

Comment:
If this is how badDay is used I don't think it's named appropriately, nilDay doesn't have the failure/negative connotation associated with bad *shrug*.